Add `get_cancelled_transaction_by_id` to FFI & update cancel callback

philipr-za opened the pull request:
## Description
This PR add a `get_cancelled_transaction_by_id(…)` function to the FFI to request a single cancelled transaction. It also updates the `callback_transaction_cancellation(…)` callback in the FFI to return a pointer to the Transaction that was cancelled.

To facilitate this the Transaction service backend was updated to allow for queries of cancelled inbound, outbound and completed transactions.

@StriderDM This PR will require updates to the Swift and JNI interfaces
@kukabi and @Jasonvdb This PR will change the return type of the transaction_cancellation callback in addition to adding the new FFI function.

## Motivation and Context
Closes https://github.com/tari-project/tari/issues/1866

## How Has This Been Tested?
Tests have been updated to exercise all the additions.

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
* [ ] Bug fix (non-breaking change which fixes an issue)
* [x] New feature (non-breaking change which adds functionality)
* [ ] Breaking change (fix or feature that would cause existing functionality to change)
* [x] Feature refactor (No new feature or functional changes, but performance or technical debt improvements)
* [ ] New Tests
* [ ] Documentation

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
* [x] I'm merging against the `development` branch.
* [x] I ran `cargo-fmt --all` before pushing.
* [ ] I have squashed my commits into a single commit.
* [ ] My change requires a change to the documentation.
* [ ] I have updated the documentation accordingly.
* [x] I have added tests to cover my changes.

StriderDM reviewed (dismissed):
Looks good, just a lint to fix.

StriderDM commented on base_layer/wallet_ffi/src/callback_handler.rs:
```diff
@@ -96,7 +96,7 @@ where TBackend: TransactionBackend + 'static
         callback_direct_send_result: unsafe extern "C" fn(TxId, bool),
         callback_store_and_forward_send_result: unsafe extern "C" fn(TxId, bool),
-        callback_transaction_cancellation: unsafe extern "C" fn(TxId),
+        callback_transaction_cancellation: unsafe extern "C" fn(*mut CompletedTransaction),
```
The header doesn't match the wallet create for this.
```
 void (*callback_transaction_cancellation)(unsigned long long)
```

needs to be changed to:
```
void (*callback_transaction_cancellation)(struct TariCompletedTransaction*),
```

StriderDM reviewed (changes requested):
=

philipr-za committed:
Add `get_cancelled_transaction_by_id` to FFI & update cancel callback

This PR add a `get_cancelled_transaction_by_id(…)` function to the FFI to request a single cancelled transaction. It also updates the `callback_transaction_cancellation(…)` callback in the FFI to return a pointer to the Transaction that was cancelled.

To facilitate this the Transaction service backend was updated to allow for queries of cancelled inbound, outbound and completed transactions.

StriderDM reviewed (approved):
LGTM
